### DISCUSSION TITLE
DS18B20 stop working randomly and will only work again after power cycle

### DISCUSSION THREAD

**Discussion by TnDA-tech** [arendst/Tasmota#15896] [Q&A] [answered]
Hi

I have around 300 devices with Tasmota and DS18B20 sensors, a mix of Shelly 1, Shelly 1PM and Sonoff TH16
In some devices after a while the temperature sensor (DS18B20) will stop showing temperature and will return to normal only after I power cycle the device. The devices can work perfect for days/weeks/months and then randomly stop showing temperature.
I couldn't find a way to fix it without power cycling the device.
I read some where that it the DS18B20 sensors might be blocking the bus and that there is no way to reset the bus without power cycling the device
I'm looking for a solution that can be done remotely because I don't always have access to the devices

Is there a way to reset the bus remotely?
Since a power cycle fixes the problem is there any command that will trigger the same action to the GPIO and might fix it remotely?

Any thoughts/ideas will be appreciated
I'll be happy to share more details if needed
Thanks

**Comment by ascillato** [comment, marked as answer]
Are you using latest Tasmota v12? If not, please update and try again.

**Reply by sfromis** [reply]
Don't know if this could help, but if you power the DS18B20 sensor from a gpio pin, you could power cycle the sensor

**Reply by TnDA-tech** [reply]
> Don't know if this could help, but if you power the DS18B20 sensor from a gpio pin, you could power cycle the sensor

Thanks @sfromis 
I do power the DS18B20 sensor from the device's GPIO 0 pin

I didn't know that I can power cycle the GPIO 
how can I do it?

**Reply by TnDA-tech** [reply]
> > Don't know if this could help, but if you power the DS18B20 sensor from a gpio pin, you could power cycle the sensor
> 
> Thanks @sfromis I do power the DS18B20 sensor from the device's GPIO 0 pin
> 
> I didn't know that I can power cycle the GPIO how can I do it?

*** CORRECTION ***
In the case of Shelly devices I power the DS18B20 from the debug/flash 3.3v output, and the sensor data is connected to GPIO 0
In Sonoff TH16 devices the power and data are from the 2.5mm jack

Is there a way to power cycle the sensor by command?

**Reply by barbudor** [reply]
What do you mean by that? Are you using parasitic power on the GPIO used for reading the DS?

What sfromis meant is having a GPIO configured as a Relay to feed the DS' VCC pin, in addition to the GPIO connected to the DS data pin.
So you can user a Power command to control it

If you are using parasitic power, it may works to configure the DS GPIO as Output_Low, restart, reconfigure back as DS and restart again
Parasitic power can be problematic especially with  non genuine DS

**Reply by TnDA-tech** [reply]
Thanks for the explanation @barbudor 

I'm not using parasitic power. the DS is connected to the device's GND, VCC (3.3v) and GPIO0 (for data)
I tried now to do as you and sfromis suggested, I configured a GPIO as a Relay and connected the DS' VCC pin to it instead of to the device's 3.3V output.

Here's what I got -
When the Relay is OFF I'm getting a temperature read
When the Relay is ON the temperature shows 0°C

it might be a good solution, I guess that by toggling the Replay GPIO I will be able to restart the DS. 
I can't do it for the existing devices as I don't have access to them, but I can try it going forward and see if it'll fix the issue when the DS stops showing temperature.

Shouldn't it be the opposite way - meaning that when the relay is ON the DS will show temperature?
Does it mean that any GPIO configured as relay can output 3.3V ?

**Reply by sfromis** [reply]
Yes, high output from any pin is 3.3V, obviously not a lot of current available. Whether ON is high or low is up to your config.

**Reply by barbudor** [reply]
> Shouldn't it be the opposite way - meaning that when the relay is ON the DS will show temperature?

If you configured as "Relay", the GPIO should be high (3V3) when ON
If you configurd as "Relay_i", it's the opposite: high (3V3) when OFF

**Reply by TnDA-tech** [reply]
Thanks @sfromis and @barbudor 
This is really helpful
I will try it going forward

**Reply by TnDA-tech** [reply]
@sfromis as you suggested I started powering the DS18B20 sensors from a gpio pin, and when the problem happened I was able to power cycle the sensor and solve it. The sensor started showing temperature again

@sfromis @barbudor  thanks for you help

**Reply by stefanbode** [reply]
Main problem is that 99% of all DS18B20 are clones with most time questionable quality. Typical you see the behavior you mentioned and also missing values if there is some electromagnetic field from motors in the nearby. The original DS18B20 cost about 10x, but they work! So it is tradeoff between reliability and cheap.

**Reply by sfromis** [reply]
As posted elsewhere in this topic: https://github.com/cpetrich/counterfeit_DS18B20/

**Comment by lalo-uy** [comment]
I have the same problem on a Sonoff Basic with a DS18B20
It works for several days and then the sensor disappears.
With extensive logging I found that Tasmota restarted just after power off
command.
It could be for an inductive load ( is a contactor ) .
Any way the sensor comes again with :
wget -q -T2 -t2 -O - http://192.168.100.5/cm?cmnd=gpio4%204

The system is version 6.7.1(sensors)  and I can't go there now to  upgrade.

The sensor is powered with 5v right from the input of the 3.3v regulator of
the Sonoff.

**Comment by rapi3** [comment]
**I have zero problems with DS18B20 ( DIP and cable-external ):**
>10 wemos D1 mini running Tasmota custom v11.1.0.3,
3x Sonof Basic running Tasmota custom v11.1.0.1
2x Sonoff RF Bridge running Tasmota custom v11.1.0.1

I had problem with one DHT11 and I blame the Chinese low quality production as faulty, long time ago I configured DHT11 power from I/O pin and I fix the HW problem with rules when it fail - no more problems.

**Reply by sfromis** [reply]
Not having trouble with DS18B20 may be down to simple luck.
https://github.com/cpetrich/counterfeit_DS18B20/

**Reply by rapi3** [reply]
True if "your DS18B20 temperature sensor is likely a fake, counterfeit, clone..." do not blame Tasmota.

**Reply by beatagerger** [reply]
oh no...
"Chinese"...
"fake"...
"Replica"...???

There can't be a better excuse???

I finally got fed up with Tasmota because either the devices could no longer be reached, or sensor data wasn't shown/transmitted, or no timer worked.

The only reason for Tasmota was the integration for Alexa (ESP32)
But since ArduinoCloud was introduced, I no longer use Tasmota. (you can't always get the built-in devices or out of the shed)

I realized that something wasn't working properly with Tasmota because I installed different systems on identical devices (ESP8266 and ESP32)...
EasyESP found all sensors immediately...

ArduinoCloud too...


NOT Tasmota!!!! Not even a "DS18b20"!!!!
